cpu/mips_pic32_common/periph/gpio.c

```diff
@@ -19,115 +19,58 @@
-#define PORTx(P)        (base_address[P].gpio[0x0])
-#define CNPUxCLR(P)     (base_address[P].gpio[0x34/0x4])
-#define CNPUxSET(P)     (base_address[P].gpio[0x38/0x4])
-#define CNPDxCLR(P)     (base_address[P].gpio[0x44/0x4])
-#define CNPDxSET(P)     (base_address[P].gpio[0x48/0x4])
-#define ODCxCLR(P)      (base_address[P].gpio[0x24/0x4])
-#define ODCxSET(P)      (base_address[P].gpio[0x28/0x4])
-#define ANSELxCLR(P)    (base_address[P].ansel[0x04/0x4])
-#define TRISxCLR(P)     (base_address[P].tris[0x04/0x4])
-#define TRISxSET(P)     (base_address[P].tris[0x08/0x4])
-
-typedef struct PIC32_GPIO_tag {
-    volatile uint32_t* gpio;
-    volatile uint32_t* ansel;
-    volatile uint32_t* tris;
-} PIC32_GPIO_T;
-
-static PIC32_GPIO_T base_address[] = {
-#ifdef _PORTA_BASE_ADDRESS
-    {
-        .gpio  = (volatile uint32_t*)_PORTA_BASE_ADDRESS,
-        .ansel = (volatile uint32_t*)&ANSELA,
-        .tris  = (volatile uint32_t*)&TRISA
-    },
-#else
-    {0 , 0, 0},
-#endif
-#ifdef _PORTB_BASE_ADDRESS
-    {
-        .gpio  = (volatile uint32_t*)_PORTB_BASE_ADDRESS,
-        .ansel = (volatile uint32_t*)&ANSELB,
-        .tris  = (volatile uint32_t*)&TRISB
-    },
-#else
-    {0 , 0, 0},
-#endif
-#ifdef _PORTC_BASE_ADDRESS
-    {
-        .gpio  = (volatile uint32_t*)_PORTC_BASE_ADDRESS,
-        .ansel = (volatile uint32_t*)&ANSELC,
-        .tris  = (volatile uint32_t*)&TRISC
-    },
-#else
-    {0 , 0, 0},
-#endif
-#ifdef _PORTD_BASE_ADDRESS
-    {
-        .gpio  = (volatile uint32_t*)_PORTD_BASE_ADDRESS,
-        .ansel = (volatile uint32_t*)&ANSELD,
-        .tris  = (volatile uint32_t*)&TRISD
-    },
-#else
-    {0 , 0, 0},
-#endif
-#ifdef _PORTE_BASE_ADDRESS
-    {
-        .gpio  = (volatile uint32_t*)_PORTE_BASE_ADDRESS,
-        .ansel = (volatile uint32_t*)&ANSELE,
-        .tris  = (volatile uint32_t*)&TRISE
-    },
-#else
-    {0 , 0, 0},
-#endif
-#ifdef _PORTF_BASE_ADDRESS
-    {
-        .gpio  = (volatile uint32_t*)_PORTF_BASE_ADDRESS,
-        .ansel = (volatile uint32_t*)&ANSELF,
-        .tris  = (volatile uint32_t*)&TRISF
-    },
-#else
-    {0 , 0, 0},
-#endif
-#ifdef _PORTG_BASE_ADDRESS
-    {
-        .gpio  = (volatile uint32_t*)_PORTG_BASE_ADDRESS,
-        .ansel = (volatile uint32_t*)&ANSELG,
-        .tris  = (volatile uint32_t*)&TRISG
-    },
-#else
-    {0 , 0, 0},
-#endif
-};
-
-static inline int check_valid_port(uint8_t port)
+#include "periph_conf.h"
+
+#define LATx(P)         ((P)[0x30/0x4])
+#define LATxCLR(P)      ((P)[0x34/0x4])
+#define LATxSET(P)      ((P)[0x38/0x4])
+#define LATxINV(P)      ((P)[0x3C/0x4])
+#define PORTx(P)        ((P)[0x20/0x4])
+#define CNPUxCLR(P)     ((P)[0x54/0x4])
+#define CNPUxSET(P)     ((P)[0x58/0x4])
+#define CNPDxCLR(P)     ((P)[0x64/0x4])
+#define CNPDxSET(P)     ((P)[0x68/0x4])
+#define ODCxCLR(P)      ((P)[0x44/0x4])
+#define ODCxSET(P)      ((P)[0x48/0x4])
+#define ANSELxCLR(P)    ((P)[0x04/0x4])
+#define TRISxCLR(P)     ((P)[0x14/0x4])
+#define TRISxSET(P)     ((P)[0x18/0x4])
```

Comment:
Using a struct for this would be a more orthodox solution, but with all those gaps I see how this is easier to handle.

Reply:
I thought about that but it went against one of the mains reasons for refactoring GPIO peripheral: avoiding storing base addresses to reduce the size of the firmware.

Reply:
If it's a compile-time constant the compiler will also optimize it away if you are using a struct.